plutus-tx/src/PlutusTx/Ratio.hs

```diff
@@ -4,80 +4,349 @@
+-- of this function, as it is significantly smaller on-chain.
+{-# INLINEABLE abs #-}
+abs :: Rational -> Rational
+abs rat@(Rational n d)
+  | n P.< P.zero = Rational (P.negate n) d
+  | P.True = rat
+
+-- | @'properFraction' r@ returns the pair @(n, f)@, such that all of the
+-- following hold:
+--
+-- * @'fromInteger' n 'P.+' f = r@;
+-- * @n@ and @f@ both have the same sign as @r@; and
+-- * @'abs' f 'P.<' 'P.one'@.
+{-# INLINEABLE properFraction #-}
+properFraction :: Rational -> (Integer, Rational)
+properFraction (Rational n d) =
+  (n `Builtins.quotientInteger` d,
+   Rational (n `Builtins.remainderInteger` d) d)
+
+-- | Gives the reciprocal of the argument; specifically, for @r 'P./='
+-- 'P.zero'@, @r 'P.*' 'recip' r = 'P.one'@.
+--
+-- = Important note
+--
+-- The reciprocal of zero is mathematically undefined; thus, @'recip' 'P.zero'@
+-- will error. Use with care.
+{-# INLINEABLE recip #-}
+recip :: Rational -> Rational
+recip (Rational n d)
+  | n P.== P.zero = Builtins.error ()
+  | n P.< P.zero = Rational (P.negate d) (P.negate n)
+  | P.True = Rational d n
+
+-- | Returns the whole-number part of its argument, dropping any leftover
+-- fractional part. More precisely, @'truncate' r = n@ where @(n, _) =
+-- 'properFraction' r@, but is much more efficient.
+{-# INLINEABLE truncate #-}
+truncate :: Rational -> Integer
+truncate (Rational n d) = n `Builtins.quotientInteger` d
+
+-- | @'round' r@ returns the nearest 'Integer' value to @r@. If @r@ is
+-- equidistant between two values, the even value will be given.
+{-# INLINEABLE round #-}
+round :: Rational -> Integer
+round x =
+  let (n, r) = properFraction x
+      m = if r P.< P.zero then n P.- P.one else n P.+ P.one
+      flag = abs r P.- half
+   in if
+          | flag P.< P.zero -> n
+          | flag P.== P.zero -> if Builtins.modInteger n 2 P.== P.zero
+                                then n
+                                else m
+          | P.True -> m
+
+-- From GHC.Real
+-- | @'gcd' x y@ is the non-negative factor of both @x@ and @y@ of which
+-- every common factor of @x@ and @y@ is also a factor; for example
+-- @'gcd' 4 2 = 2@, @'gcd' (-4) 6 = 2@, @'gcd' 0 4@ = @4@. @'gcd' 0 0@ = @0@.
+{-# INLINABLE gcd #-}
+gcd :: Integer -> Integer -> Integer
+gcd a b = gcd' (P.abs a) (P.abs b) where
+    gcd' a' b'
+        | b' P.== P.zero = a'
+        | P.True         = gcd' b' (a' `Builtins.remainderInteger` b')
+
+-- From GHC.Real
+-- | Given a numerator and denominator, produces a 'Rational' by dividing both
+-- numerator and denominator by their greatest common divisor.
+{-# INLINABLE reduce #-}
+reduce :: Integer -> Integer -> Rational
+reduce x y
+    | y P.== 0 = P.traceError P.ratioHasZeroDenominatorError
+    | P.True     =
+        let d = gcd x y in
+          Rational (x `Builtins.quotientInteger` d)
+                   (y `Builtins.quotientInteger` d)
+
+-- Helpers
+
+-- Euclid's algorithm
```

Comment:
This is a bit confusing because we have two versions of the Euclidean algortihm: `euclid` here (which uses `modInteger`) and `gcd'` in `gcd`, which uses `remainderInteger`.  Can we get away with just one of these?  They behave differently when negative numbers are involved and it's hard to work out what's going on.  GHC.Real uses something the same as `ghc`, but I have to admit that `euclid` seems a bit better behaved.

Reply:
It also seems that `euclid` is always called with a positive second argument (in which case it always returns a positive result), except perhaps if we're scaling by a negative number in the `Module` instance, which we probably don't care too much about anyway.  That might be a useful thing to mention.

Reply:
I originally didn't expose `gcd`, but I was asked to put it back in by @michaelpj. I honestly agree that exposing this internal detail is unnecessary, and adds more code (and confusion!), and would prefer to have it gone.

Reply:
Computing `gcd` is independently useful in some cases, it's not *just* an implementation detail of `Rational`.